keps/sig-auth/0000-authority-delegation.md

```diff
@@ -0,0 +1,259 @@
+delegation in Kubernetes.
+
+## Motivation
+
+Kubernetes has no mechanism to propagate authority through the stack.
+
+* For use in authorization: All controllers act as deputies but we only have
+  quality control over the core controllers. 3rd party controllers may turn out
+  to be more easily confused and tricked into using their authority incorrectly,
+  putting our users at unnecessary risk. Ideally the authority of a controller
+  would be limited to (a subset of) the authority of the users actively using
+  the controller.
+* For use in audit: Lack of provenance of authority greatly restricts what we
+  can implement as far as audit logging and access transparency. Tracing the
+  authority of a low level action to the source is painful, imprecise,
+  dependent on k8s version, even when each actor in the path is cooperative.
+
+The proposed authorization model seeks to solve these issues.
+
+<!---
+
+## Table of Contents
+
+## Summary
+
+
+### Goals
+
+### Non-Goals
+
+-->
+
+
+## Proposal
+
+### Selected Authority
+
+Today Kubernetes supports a kind of ambient authority. However support for
+selected authority is a generally useful model to support.
+
+This is particularly useful in a deputy scenario to prevent confused deputies.
+When a controller acts on behalf of an end principal, the authority by which it
+acts should be explicit. This prevents a controller from misusing permission
+granted ambiently to the controller.
+
+Authority is selected per request by passing an “Authority-Selector” header
+with the request:
+
+```
+Authority-Selector: {“authorityReference”:{“kind”:”ClusterRole”,”name”:”viewer”}}
+```
+or:
+```
+Authority-Selector: {“authorityReference”:{“kind”:”AuthorityDelegation”,”uid”:”0000-0...”}}
+```
+
+The value of the selector is a `SelectedAuthority` object in the
+`authorization.k8s.io` API group:
+
+```go
+type SelectedAuthority struct {
+  metav1.TypeMeta
+  AuthorityReference AuthorityReference
+}
+
+type AuthorityReference struct {
+  metav1.TypeMeta
+  Name string
+  UID types.UID
+}
+```
+
+When a request is received by an API, the API would normally iterate over the
+set of authorizers and each authorizer would check all policies that would
+permit the action, searching for a policy that permits the action and is bound
+to the user.
+
+When a request is received by the API with a selected authority, the API would
+check the specific authorizer that supports the AuthorityReference kind, and
+the authorizer would check the request against the specific policy referenced
+in by the SelectedAuthority.
+
+### Persistent Authority Delegation
+
+#### AuthorityDelegation
+
+A new non-namespaced persistent object will be added to the
+`authorization.k8s.io` API group to store instances of authority delegation.
+
+```go
+type AuthoritiyDelegation struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // The user that created the delegation.
+  User Info
```

Comment:
Is this just for audit purposes? If that's the case, can we just rely on the audit logs?

Reply:
It's for audit purposes and also allows an admin to list all persisted AuthorityDelegations and see which historic delegations might still be granting permission. This is useful, for example, when a team member leaves a team and is removed from ACL groups. You don't necessarily want all jobs they started to stop immediately, but you also want to understand what remnants of their authority still persist.

As far as audit, I would prefer to hoist the delegation history and terminal authority (i.e. Subject) into the audit log so that the complete authority is recorded in each audit log and users don't have to analyze a graph of logs to understand why some access was allowed.